Refactor work queues

edwintorok opened the pull request:
Currently XAPI has a single thread that processes all events from xenopsd. This can lead to a significant delay between the VM running, and XAPI reporting its state as running, e.g. you can receive a UDP packet from the VM, but XAPI still doesn't report it as running for minutes (CA-87377).

The [fix for that](
https://github.com/edwintorok/xen-api/commit/b367bf86d3af4f773db9bf5d1500a4dec0f99bfa#diff-344dc1d17c4663add7fe5500813feef2) parallelizes event processing on a per-VM basis.

To avoid duplicating code the `WorkerPool` from `xenopsd` should be shared between `xenopsd` and `xapi`. In order to do that we should remove any `xenopsd` specific details from it.
Initially I thought of doing that with an `'a Item.t` , but there were multiple operations that had to be performed on it, essentialy ending up with a record threaded through everything, like a home grown functor. So I turned this into into a proper functor (see [its documentation here](https://github.com/xapi-project/xenopsd/pull/337/files#diff-08ea6b223a2a8da3d28bac7c572e32a2)):
 * the work items submitted should be abstract, the WorkerPool shouldn't have any knowledge of its internals (`Item.t`)
 * the item to be executed should be as generic as possible, i.e. it should be able to pass just a `unit -> unit` to be executed by the workers, all the [`xenopsd` specific task execution](https://github.com/xapi-project/xenopsd/pull/337/commits/54ae6d6f5c8d171bdb438bbc9b21f977cc81e22a#diff-469ce4709f9a70be7fae148c604342d1L612) is moved to `execute` and `finally`.
* the semantics of `execute` and `finally` are documented in the interface (in short: `execute` and `finally` are always called on an `Item.t` to run them, exceptions are logged and ignored)
* `should_keep` is specific to xenopsd, used for removing certain tasks if they are already present in the queue
* there are some functions used only by debugging/diagnostics, e.g. `describe_item`, `describe_task`.

Most of the internal functions of the redirector and worker pool are hidden, except:
 * setting the size of the worker pool (which you must do in order to have the jobs execute at all)
 * submitting jobs to the pool

There are 2 global queues in xenopsd (`default` and `parallel`), which would've been nice to avoid, however there is more global state in the redirector (the `overrides` map), which would've made this PR even larger.

The scheduling algorithm is briefly explained [here](https://github.com/xapi-project/xenopsd/pull/337/files#diff-08ea6b223a2a8da3d28bac7c572e32a2R68), in the case of `xapi` the  tags will be the VM ids for example.

This PR refactors xenopsd to extract the worker pool as a separate file, independent of `xenopsd`-specific implementation details, i.e. as a functor. The functionality and scheduling algorithm should be unaffected by the refactoring, this PR should be a no-op in terms of runtime visible change, except tiny differences in the debug output.
I've also added a new test to exercise the new worker pool code paths.

A second PR will move the code (`xapi_work_queues.ml{,i}` and `xwq_test.ml`) to a separate repository `xapi-work-queues` and update `xenopsd` and `xen-api` to use that.

The `xen-api` instance of the worker pool functor will look like this:
https://github.com/edwintorok/xen-api/commit/8efe9e0e136b9e3b10738668eb89c0c0295c9184#diff-344dc1d17c4663add7fe5500813feef2R1931

I tried to keep commits that functorize (or refactor in preparation for that) separate from the commits that just move code around to make it easier to follow.

jonludlam commented:
Why do you want to remove Rpc.t from the interface? It's not particularly for debugging, but also ensures we don't end up embedding Xml in JSON or vice-versa.

edwintorok commented:
There isn't a meaningful way to dump the `unit -> unit`, it looked a bit awkward to construct a `Rpc.rpc_of_unit ()`, thought `Rpc.t` is just an implementation detail on how things are dumped and shouldn't be exposed. But you do have a point about mixing Xml and JSON, I'll put it back. (and make operation dumping use `Rpc.t` too).

thomassa commented:
@jonludlam You've commented on this pull-request; have you reviewed all of it and made all the comments you want, or does it still need review? (Or in immediate practical terms: should I look at it?)

edwintorok commented:
Pushed updated tests (for round-robin), refactored to remove the globals from the interface, and updated the docs. See also https://edwintorok.github.io/xapi-work-queues/Xapi_work_queues.html
Please let me know if I missed anything.

edwintorok commented:
Rebased on top of master, squashed the fixups, and checked that the unit tests still pass.

lindig commented:
I know that this was reviewed by @jonludlam working with @edwintorok. Is this ready to be merged or is there work left to be done?

edwintorok commented:
There were some small changes left to be done, e.g. refactoring how `should_keep` works to take a list and return a list, separate testing of internal vs external API and adding a test to ensure that if one queue gets flooded with tasks it doesn't affect the processing of events from other queues. 
I'll try to find some time to do them and then ping this PR.

mseri commented:
:+1: unflag it when it's ready

mseri commented:
@edwintorok this needs rebasing.
@jonludlam any update on the review?

edwintorok commented:
We've reviewed the interface and the refactoring of the worker queue code itself with @jonludlam, which looked good.
I've done the suggested cleanups as fixup commits on top (and rebased on master again due to merge conflict).

We agreed that the remaining cleanup commits can be reviewed by anyone:

@edwintorok
fixup! refactor work queues: move to another dir
867e753
@edwintorok
fixup! refactor work queues: drop dead code
5b3d454
@edwintorok
fixup! refactor work queues: drop finally from the interface
69449e6
@edwintorok
refactor work queues: use xapi-stdext-threads instead of stdext
aaa96ee
edwintorok added some commits 40 minutes ago
@edwintorok
refactor work queues: port to alcotest
ee0494e
@edwintorok
refactor work queues: fix warnings
d9e3e7c

edwintorok commented:
I've restricted the reindentation to just the file touched by this PR to avoid conflicts with other feature branches.
FWIW the commands I used were roughly these (note that interactive rebase cancels out the ignore-space-change, so can't do both at same time)
```
git checkout master
git pull upstream master
ocp-indent -i lib/xenops_server.ml
git commit -sam 'refactor work queues: reindent'
git tag reindent
git checkout refactor-work-queues
git rebase -Xignore-space-change --no-autosquash master
git filter-branch -f --tree-filter "git diff-tree --name-only --diff-filter=AM -r --no-commit-id \$GIT_COMMIT| grep '.*ml\|.*mli'|xargs ocp-indent -i" reindent..HEAD
git rebase -i master
```

edwintorok commented:
I did a test merge this branch with `qemu-upstream`, `feature/qemu-datapath` and `sr-iov`, the confllicts are in files not touched by this PR:
```
build@5fef6e845c94:~/local/repos/xenopsd %  git merge upstream/sr-iov
Merge made by the 'recursive' strategy.
 xc/device.ml                      |  62 ++++++++++-
 xc/device.mli                     |  24 ++++-
 xc/device_common.ml               |  21 +++-
 xc/device_common.mli              |   2 +-
 xc/domain.ml                      |   9 +-
 xc/hotplug.ml                     |   3 +-
 xc/jbuild                         |   1 +
 xc/xenops_server_xen.ml           | 802 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------------------------
 xl/xenops_server_xenlight.cppo.ml |   5 +-
 9 files changed, 532 insertions(+), 397 deletions(-)
build@5fef6e845c94:~/local/repos/xenopsd %  git merge upstream/feature/qemu-datapath
Auto-merging xc/xenops_server_xen.ml
Auto-merging xc/jbuild
Auto-merging xc/hotplug.ml
Auto-merging xc/device_common.ml
Auto-merging xc/device.ml
Merge made by the 'recursive' strategy.
 scripts/xen-backend.rules.in |  2 +-
 xapi-xenopsd-xc.opam         |  1 +
 xc/device.ml                 | 21 +++++++++++++++++++--
 xc/device_common.ml          |  2 +-
 xc/hotplug.ml                | 18 ++++++++++++------
 xc/jbuild                    |  1 +
 xc/xenops_server_xen.ml      | 54 ++++++++++++++++++++++++++++++++++++++++++++++++++++--
 xl/device_common.ml          |  2 +-
 8 files changed, 88 insertions(+), 13 deletions(-)
build@5fef6e845c94:~/local/repos/xenopsd %  git merge upstream/qemu-upstream
Auto-merging xl/xenops_server_xenlight.cppo.ml
Auto-merging xl/device_common.ml
Auto-merging xc/xenops_server_xen.ml
CONFLICT (content): Merge conflict in xc/xenops_server_xen.ml
Auto-merging xc/domain.ml
Auto-merging xc/device_common.mli
Auto-merging xc/device_common.ml
CONFLICT (content): Merge conflict in xc/device_common.ml
Auto-merging xc/device.mli
Auto-merging xc/device.ml
Auto-merging lib/xenops_utils.ml
Automatic merge failed; fix conflicts and then commit the result.
```

mseri commented:
I am happy to merge. @lindig do you have any objections?

lindig commented:
Merge it. But we really need a better story about working on feature branches. We can't constantly have conflicts are risky to resolve.

edwintorok commented:
Rebased the fixups, thanks for the reviews!

mseri commented on xapi-work-queues/test/jbuild:
```diff
@@ -0,0 +1,44 @@
+(executable (
+  (name main)
+  (libraries (alcotest
+              xapi-stdext-pervasives
```
This did also need `xapi-stdext-threads`
